crates/sui-core/src/execution_engine.rs

```diff
@@ -237,3 +251,72 @@ fn transfer<S>(
     temporary_store.write_object(object);
     Ok(())
 }
+
+/// Transfer the gas object (which is a SUI coin object) with an optional `amount`.
+/// If `amount` is specified, the gas object remains in the original owner, but a new SUI coin
+/// is created with `amount` balance and is transferred to `recipient`;
+/// if `amount` is not specified, the entire object will be transferred to `recipient`.
+/// `tx_ctx` is needed to create new object ID for the split coin.
+/// We make sure that the gas object's version is not incremented after this function call, because
+/// when we charge gas latter, its version will be officially incremented.
+fn transfer_sui<S>(
+    temporary_store: &mut AuthorityTemporaryStore<S>,
+    mut object: Object,
+    recipient: SuiAddress,
+    amount: Option<u64>,
+    tx_ctx: &mut TxContext,
+) -> SuiResult {
+    #[cfg(debug_assertions)]
+    let version = object.version();
+    if let Some(amount) = amount {
+        let mut gas_coin = GasCoin::try_from(&object)?;
+        let balance = gas_coin.value();
+        fp_ensure!(
+            balance >= amount,
+            SuiError::TransferInsufficientBalance {
+                balance,
+                required: amount,
+            }
+        );
+        let new_object = Object::new_move(
+            MoveObject::new(
+                GasCoin::type_(),
+                bcs::to_bytes(&GasCoin::new(
+                    tx_ctx.fresh_id(),
+                    OBJECT_START_VERSION,
+                    amount,
+                ))
+                .expect("Serializing gas object cannot fail"),
+            ),
+            Owner::AddressOwner(recipient),
+            tx_ctx.digest(),
+        );
+        temporary_store.write_object(new_object);
+
+        // This is necessary for the temporary store to know this new object is not unwrapped.
+        let newly_generated_ids = tx_ctx.recreate_all_ids();
+        temporary_store.set_create_object_ids(newly_generated_ids);
+
+        // Safe because we have checked that balance covers amount.
+        gas_coin.0.balance.withdraw(amount);
+        let move_object = object
+            .data
+            .try_as_move_mut()
+            .expect("Gas object must be Move object");
+        move_object.update_contents(
+            bcs::to_bytes(&gas_coin).expect("Serializing gas coin can never fail"),
+        );
+        // update_contents increments object version, reverts it.
+        move_object.revert_version_increase();
```

Comment:
It seems like we have two different patterns for avoiding double version bumps (stop callee from doing a version bump vs do double version bump and then revert). If we can, I think it would be nice to only use one pattern if we can for consistency. 

Exposing a version of `update_contents` that does not do a version bump seems like the path of least resistance, but am not sure. Going this route would also let us eliminate `revert_version_increase` (which is only used here).